src/getPublicKeys.c

```diff
@@ -0,0 +1,265 @@
+static void returnOneKey(bool isSingleKeyExport)
+{
+	ctx->responseReadyMagic = 0;
+
+	// Check security policy
+	security_policy_t policy = (isSingleKeyExport) ?
+	                           policyForGetExtendedPublicKey(&ctx->pathSpec) :
+	                           policyForGetExtendedPublicKeyBulkExport(&ctx->pathSpec);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// Calculation
+		deriveExtendedPublicKey(
+		        & ctx->pathSpec,
+		        & ctx->extPubKey
+		);
+		ctx->responseReadyMagic = RESPONSE_READY_MAGIC;
+	}
+
+	switch (policy) {
+#	define  CASE(policy, step) case policy: {ctx->ui_step = step; break;}
+		CASE(POLICY_PROMPT_WARN_UNUSUAL,    GET_KEY_UI_STEP_WARNING);
+		CASE(POLICY_PROMPT_BEFORE_RESPONSE, GET_KEY_UI_STEP_DISPLAY);
+		CASE(POLICY_ALLOW_WITHOUT_PROMPT,   GET_KEY_UI_STEP_RESPOND);
+#	undef   CASE
+	default:
+		ASSERT(false);
+	}
+	getPublicKeys_getOneKey_ui_runStep();
+}
+
+// ============================== INIT ==============================
+
+enum {
+	HANDLE_INIT_UI_STEP_CONFIRM = 100,
+	HANDLE_INIT_UI_STEP_RESPOND,
+	HANDLE_INIT_UI_STEP_INVALID,
+} ;
+
+static void getPublicKeys_handleInit_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = getPublicKeys_handleInit_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+	UI_STEP(HANDLE_INIT_UI_STEP_CONFIRM) {
+		char secondLine[100];
+		explicit_bzero(secondLine, SIZEOF(secondLine));
+		snprintf(secondLine, SIZEOF(secondLine), "%u public keys?", ctx->remainingPaths);
+		ASSERT(strlen(secondLine) < SIZEOF(secondLine));
+
+		ui_displayPrompt(
+		        "Confirm export",
+		        secondLine,
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_INIT_UI_STEP_RESPOND) {
+		ASSERT(ctx->remainingPaths > 0);
+		returnOneKey(false); // runs another UI state machine
+
+		ctx->stage = GET_KEYS_STAGE_GET_KEYS;
+	}
+	UI_STEP_END(HANDLE_INIT_UI_STEP_INVALID);
+}
+
+static void getPublicKeys_handleInitAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		CHECK_STAGE(GET_KEYS_STAGE_INIT);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		// parse data
+
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		parsePath(&view);
+
+		size_t remaining = view_remainingSize(&view);
+		switch (remaining) {
+		case 0: {
+			TRACE();
+			ctx->remainingPaths = 0;
+			break;
+		}
+		case 4: {
+			// read the number of remaining keys
+			uint32_t remainingPaths = parse_u4be(&view);
+			ASSERT(view_remainingSize(&view) == 0);
+			TRACE("num keys: %d", remainingPaths);
+			VALIDATE(remainingPaths <= MAX_PUBLIC_KEYS, ERR_INVALID_DATA);
+
+			ASSERT_TYPE(ctx->remainingPaths, uint16_t);
+			ASSERT(remainingPaths <= UINT16_MAX);
+			ctx->remainingPaths = (uint16_t) remainingPaths;
+			break;
+		}
+		default: {
+			THROW(ERR_INVALID_DATA);
+		}
+		}
+	}
+
+	if (ctx->remainingPaths == 0) {
```

Comment:
why do we have this fast-track? Especially because it is skipping the policy below...

Reply:
it skips the prompt for the count of exported keys, but given that `policyForGetPublicKeysInit` takes as a parameter the remaining paths count, it probably makes more sense to move the check for zero there and return POLICY_ALLOW_WITHOUT_PROMPT in that case

Reply:
ahh, `returnOneKey` returns the key already (something that would be done for the second time with `getPublicKeys_handleInit_ui_runStep`

Reply:
but it seems to me that we should be able to get rid of this fast-track, if for `remainingPaths == 0` we had `POLICY_ALLOW_WITHOUT_PROMPT` policy here, we would keep in ctx apart from "remainingPaths" also the "totalPaths" count which would be the count of total keys to be returned from this flow (i.e. won't decrease throughout the flow) and then the `returnOneKey's` policy would be `POLICY_PROMPT_BEFORE_RESPONSE` if `totalPaths` equals zero, and there would be no need to pass any bool or policy to returnOne key as all the info needed to evaluate the policy (i.e. whether the flow is a single or bulk key export, based on the totalPaths value) would live within the context